Add Support for Multiple l in Steinhardt

b-butler opened the pull request:
<!-- Provide a general summary of your changes in the Title above. -->

## Description
<!-- Describe your changes in detail. -->
This adds support for multiple `l` values in the `Steinhardt` class. I also introduce significant speed ups through three changes
- Only construct `fsph::PointSPHEvaluator` once per thread per `baseCompute` (the construction is somewhat costly)
- Reduce allocations of `std::vector` this surprisingly comprised a decent bit of the time computing.
- Reduce calls to `ManagedArray::getIndex` or the use of the `operator()`. This reduces `std::vector` allocations further and reduces computation by accessing the data directly through `operator[]`.

The changes are implemented in a backwards compatible way. Currently if the user only provides a single `l` value, everything behaves as before, but if they provide multiple, we return a list of parameters for all the order parameters provided by the compute.

Edit: The new approach uses a multidimensional `ManagedArray` and NumPy arrays.

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- Replace ??? with the issue number that this pull request resolves, if applicable. -->
Resolves: #747

This is desired since using multiple `Steinhardt` objects is significantly slower than using `LocalDescriptors` and a custom Numba solution to compute the Steinhardt order parameters (even in serial).

Here is a brief benchmark comparing the new code to the old and the custom code mentioned in issue #747. For more details see the comment below.

| Name | Time| Commit |
| ------- | ------- | -------- |
| New Code (l = [0, 12]) | ~169ms | 345a0dd |
| New Code with `ManagedArray`s | ~330ms | 8371e9a |
| Custom Code (l = [0, 12]) | ~334ms | NA |
| New Code (l=6) | ~57.7ms | 345a0dd |
| Old Code (l=6) | ~190ms | f9599480 |

\* The commit hash for "Old Code" may be wrong. I just used the HEAD of the master branch for comparison.

Edit: Due to rebasing the new code commit hash does not exist on the branch any more.

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to
     see how your changes affect other areas of the code, etc. -->
I added a test like the one that checks the validity of the spherical harmonic computation using multiple `l` values on a single compute.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:
<!-- Put an `x` in all the boxes that apply. If you're unsure about any of
     these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

b-butler committed:
Allow Steinhardt to compute for multiple l values

- Also changes SolidLiquid to work with the new Steinhardt classes
- The Python interface does not have any breaking changes as if only one
l value is chosen then lists are not returned only the first item in the
list.

b-butler committed:
Move construction of fsph::PointSPHEvaluator outside loop

- This provides a significant speed up ~2x (on my local computer) from
the previous performance.

b-butler committed:
Steinhardt: only allocate Ylms once per baseCompute

b-butler committed:
Reduce calls to ManagedArray<>::{getIndex,operator()}

* Improves performance

b-butler committed:
Add documentation of change

b-butler committed:
Add tests for multiple l in Steinhardt

b-butler committed:
Update Changelog

b-butler committed:
Update credits

b-butler committed:
Apply clang-format-10 to Steinhardt.{cc,h}

clang-format-6 did not work

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -90,15 +99,37 @@ void Steinhardt::compute(const freud::locality::NeighborList* nlist,
+            auto max_l = std::max_element(m_ls.begin(), m_ls.end());
+            if (max_l == m_ls.end())
+            {
+                throw std::invalid_argument("Selected l's do not have a maximum.");
```
Is this code path reachable? I don't think this is a case we need to catch. We would be able to simplify the code if this was removed (make `max_l` `const auto` and dereference the pointer immediately).

b-butler replied:
Only if they pass an empty list, which we can check for elsewhere.

bdice replied:
That should raise an error in the constructor. Can't believe I didn't think of the empty case, I stared at that for a while. 🙃

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,80 +237,113 @@ void Steinhardt::computeAve(const freud::locality::NeighborList* nlist,
+                        {
+                            // Adding all the qlm of the neighbors. We use the
+                            // vector function signature for indexing into the
+                            // arrays for speed.
```
Would this be better to cache the base index and re-use `base_index + k` for all `k`? That pattern is used in a few other places that were changed for this PR.
_edit: I made this change in my branch with other suggestions_

b-butler replied:
Yes it would be better, I focused on `baseCompute`, but I should extend that focus.

bdice committed:
Add missing include.

bdice commented on freud/order.pyx:
```diff
@@ -543,11 +544,12 @@ cdef class Steinhardt(_PairCompute):
         a finite number of neighbors. If you would like to ignore this
         distinction, you can mask the output order parameter values using
-        NumPy: :code:`numpy.nan_to_num(particle_order)`.
+        NumPy: :code:`numpy.nan_to_num(particle_order[0])`.
```
I think the `[0]` is not helpful here -- it's assuming you have an array named `particle_order` and this `[0]` only applies if users have multiple l values.
```suggestion
        NumPy: :code:`numpy.nan_to_num(particle_order)`.
```

bdice commented on freud/order.pyx:
```diff
@@ -543,11 +544,12 @@ cdef class Steinhardt(_PairCompute):
     Args:
-        l (unsigned int):
-            Spherical harmonic quantum number l.
+        l (unsigned int or list of unsigned int):
```
```suggestion
        l (unsigned int or sequence of unsigned int):
```

bdice commented on freud/order.pyx:
```diff
@@ -543,11 +544,12 @@ cdef class Steinhardt(_PairCompute):
-            Spherical harmonic quantum number l.
+        l (unsigned int or list of unsigned int):
+            One or more spherical harmonic quantum number l's to compute the
+            Steinhardt order parameter for.
```
Avoid ending with a preposition.
```suggestion
            One or more spherical harmonic quantum number l's used to compute
            the Steinhardt order parameter.
```

bdice commented on freud/order.pyx:
```diff
@@ -568,6 +570,8 @@ cdef class Steinhardt(_PairCompute):
 
     def __cinit__(self, l, average=False, wl=False, weighted=False,
                   wl_normalize=False):
+        if not hasattr(l, "__len__"):
```
Use `isinstance(l, collections.abc.Sequence)`.

b-butler replied:
For some reason at the time I was trying to avoid the import, but will do.

bdice commented on freud/order.pyx:
```diff
@@ -598,41 +602,64 @@ cdef class Steinhardt(_PairCompute):
+        ls = self.thisptr.getL()
+        if ls.size() == 1:
+            return ls[0]
+        return ls
```
```suggestion
        return ls[0] if ls.size() == 1 else ls
```

bdice commented on freud/order.pyx:
```diff
@@ -598,41 +602,64 @@ cdef class Steinhardt(_PairCompute):
+        order = self.thisptr.getOrder()
+        if order.size() == 1:
+            return order[0]
+        return order
```
```suggestion
        return order[0] if order.size() == 1 else order
```

bdice commented on freud/order.pyx:
```diff
@@ -598,41 +602,64 @@ cdef class Steinhardt(_PairCompute):
+                &order_arrays[0], freud.util.arr_type_t.FLOAT)
+        return [freud.util.make_managed_numpy_array(&order_arrays[i],
+                                                    freud.util.arr_type_t.FLOAT)
+                for i in range(order_arrays.size())]
```
Simplify this using a similar construction to the suggestions above. Also applies below. Avoid repeating the complicated syntax of `freud.util.make_managed_numpy_array`.

bdice commented on freud/order.pyx:
```diff
@@ -598,41 +602,64 @@ cdef class Steinhardt(_PairCompute):
+        """:math:`\\left(N_{particles}\\right)` :class:`numpy.ndarray` or list
+        of `numpy.ndarray`: Variant of the Steinhardt order parameter for each
+        particle (filled with :code:`nan` for particles with no neighbors).
+        Is a list of NumPy arrays if more than one ``l`` was specified."""
```
**Design question:** As a user, I would prefer to get higher-dimensional NumPy arrays instead of lists-of-arrays (or vectors of ManagedArrays in C++) in any place where the dimensions match (i.e. not with a size that varies with `l`, of course).

b-butler replied:
I thought about this, but this was the easiest thing to do first pass. I think I can modify it to have that behavior without many changes though.

bdice committed:
Improvements in Steinhardt.cc. Avoid "auto" when size_t is needed (reduces warnings in MSVC). Use "auto" and "const" in more places.

bdice committed:
Edit comment, use const auto&.

bdice reviewed (changes requested):
Some suggested changes here: https://github.com/glotzerlab/freud/compare/feature/multi-sph-l...feature/multi-sph-l-bdice

Please share whatever numbers/benchmarks you have handy. No need to do a _ton_ of work for that, but it would be nice to know how much of a performance difference there is for a system with 4000 particles and `l = [4, 5, 6, 7, 8, 9, 10, 11, 12]`.

Please re-request my review once you've had a chance to merge in my suggested changes and address comments. Thanks!

bdice commented on freud/_order.pxd:
```diff
@@ -70,20 +70,20 @@ cdef extern from "HexaticTranslational.h" namespace "freud::order":
-        const freud.util.ManagedArray[float] &getParticleOrder() const
-        float getOrder() const
+        const vector[freud.util.ManagedArray[float]] &getQl() const
+        const vector[freud.util.ManagedArray[float complex]] &getQlm() const
```
This is making our CI sad. Is there a compatibility issue with that version of Cython, or something weird with this syntax? Might try a typedef of `float complex` that is a single symbol, for Cython to recognize this syntax...
```
Error compiling Cython file:
------------------------------------------------------------
...
        unsigned int getNP() const
        void compute(const freud._locality.NeighborList*,
                     const freud._locality.NeighborQuery*,
                     freud._locality.QueryArgs) except +
        const vector[freud.util.ManagedArray[float]] &getQl() const
        const vector[freud.util.ManagedArray[float complex]] &getQlm() const
                                                  ^
------------------------------------------------------------

/root/ci/freud/freud/_order.pxd:79:51: Expected ']', found 'complex'
```

b-butler replied:
I am not sure, it passes on some tests, and locally, but not for some :man_shrugging: ? But I can typedef and see if that helps.

Edit: use emoji rather than text that formats weirdly.

b-butler committed:
Require at least one l to be specified for Steinhardt

Also moves the check to Cython/Python.

b-butler committed:
Apply suggestions from code review

Co-authored-by: Bradley Dice <bdice@bradleydice.com>

b-butler committed:
Move ManagedArray::getIndex outside for loop

b-butler committed:
Use typedef instead of float complex

b-butler committed:
Use a typedef for float complex

b-butler committed:
Simplify Steinhardt properties

Always create array lists and afterwards check for length.

b-butler committed:
Explicitly convert vector to list for Cython compatibility

b-butler committed:
Apply clang-tidy and cppcheck suggestions

b-butler commented:
Some rough bench-marking on a FCC system of 16,000 particles with some noise. The benchmarks were run on my laptop which has 4 cores with hyperthreading (an i5-8250U for those interested). I did not bother to close all my programs to get the cleanest reading. I ran the benchmarks in a Jupyter notebook with the `%%timeit` magic command a few times and made sure they were sufficiently close. All neighborlist were precomputed to prevent this from having an effect. This is for the standard computation (not averaged or `Wl`). 

| Name | Time| Commit |
| ------- | ------- | -------- |
| New Code (l = [0, 12]) | ~169ms | 345a0dd |
| New Code with `ManagedArray`s | ~330ms | 8371e9a |
| Custom Code (l = [0, 12]) | ~334ms | NA |
| New Code (l=6) | ~57.7ms | 345a0dd |
| Old Code (l=6) | ~190ms | f9599480 |

I am going to put this at the top of the PR but wanted to record when I added it as well.

Edit: Added with`ManagedArray` and commit hashes. The one for the old code may not be correct, I used the current HEAD on master for comparison.

bdice commented on freud/order.pyx:
```diff
@@ -598,41 +605,59 @@ cdef class Steinhardt(_PairCompute):
+        ls = self.thisptr.getL()
+        # list conversion is necessarily explicit as otherwise CI Cython
+        # complains about compiling the expression with two different types.
+        return ls[0] if ls.size() == 1 else list(ls)
```
Or just dodge the Cython confusion entirely by pre-emptively creating a list. Also applies below.
```suggestion
        ls = list(self.thisptr.getL())
        return ls[0] if len(ls) == 1 else ls
```

b-butler committed:
Use ManagedArrays over vectors whenever possible

Makes quantities that don't depend on m (Steinhardt) a unified array for
all l.

b-butler commented:
I changed to using `ManagedArray` objects over vectors of `ManagedArray` objects. I did notice it was slower about the same speed for the custom code benchmarked above and mentioned in #747.

bdice commented:
> I changed to using `ManagedArray` objects over vectors of `ManagedArray` objects. I did notice it was slower about the same speed for the custom code benchmarked above and mentioned in #747.

We don’t have to make this change if there is a measurable performance difference. I am a little surprised, however. Can you share your benchmarks with commit ids for each version tested? I can review the implementation and see what might be wrong with the performance if you agree that this design choice (one large array) is preferable.

b-butler commented:
@bdice I added commit hashes. I may not get back to this today to fix the tests, but it works locally for me, and feel free to test our the commits locally to see if you find what might be causing the slow down. I think the singular NumPy array is nice if we can do it without sacrificing significant performance.

vyasr commented:
I opted for simplicity, generality, and safety in the implementation of `ManagedArray` over performance wherever there was an obvious choice to be made. In code where we have higher dimensional arrays (like PMFTs) or in general in any code where we index into ManagedArray objects in nested loops, those choices have noticeable performance penalties. I've worked around these before, but I can work with @b-butler to resolve this more fundamentally if we think it's worthwhile. There will by necessity be some tradeoffs to be made if we choose that, at the very least in the form of more code to maintain if we try to maintain both slow/safe and fast/unsafe code paths, but if the bottlenecks are hindering analysis then maybe it's worth it.

b-butler commented:
@vyasr  I think this was the right choice especially since generally we don't need this high performance multidimensional access. I am not sure if this case would warrant creating unsafe paths for performance reasons, but if you have thoughts on improvements on my usage please suggest them.

b-butler committed:
Simplify properties in Steinhardt with early list conversion

b-butler committed:
Add support for legends in histogram with multiple datasets

b-butler commented:
@vyasr and @bdice what are your thoughts on what we should do from here with both working versions.

bdice commented:
I see the same performance between commit 345a0dd7, which used a vector of ManagedArrays, and commit 13e856a, which uses a 2D Managed Array. Both commits are about 2x faster than the `master` branch for large systems and one value of `l`.

345a0dd7:
![image](https://user-images.githubusercontent.com/3943761/120041250-8970bd00-bfcd-11eb-9f55-b483429cafab.png)

13e856a:
![image](https://user-images.githubusercontent.com/3943761/120041283-955c7f00-bfcd-11eb-8029-d287e4faddef.png)

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,80 +234,117 @@ void Steinhardt::computeAve(const freud::locality::NeighborList* nlist,
+                        auto& qlmi = m_qlmi[l_index];
+                        const auto ave_index = qlmiAve.getIndex({static_cast<unsigned int>(i), 0});
+                        const auto nb2_index = qlmi.getIndex({nb2.point_idx, 0});
+                        for (unsigned int k = 0; k < m_num_ms[l_index]; ++k)
```
Double-check that we use `size_t` for indexing.
```suggestion
                        for (size_t k = 0; k < m_num_ms[l_index]; ++k)
```

b-butler replied:
We do, I think I got confused with something converting to `unsigned int` rather than `size_t`.

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,80 +234,117 @@ void Steinhardt::computeAve(const freud::locality::NeighborList* nlist,
+                const size_t index = qlmiAve.getIndex({static_cast<unsigned int>(i), 0});
+                const size_t qliAve_index = qliAve_i_start + l_index;
+
+                for (unsigned int k = 0; k < m_num_ms[l_index]; ++k)
```
```suggestion
                for (size_t k = 0; k < m_num_ms[l_index]; ++k)
```

bdice reviewed (approved):
A couple minor suggestions, otherwise LGTM. I read through this kind of quickly but I think it's fine, the diffs are just messy with the additional level of indentation. We have pretty good tests for this class now, so I feel relatively comfortable with the changes.

b-butler committed:
Use size_t for for loop indices

Co-authored-by: Bradley Dice <bdice@bradleydice.com>

b-butler committed:
Use size_t more consistently

b-butler committed:
Add comment as to pre-allocation of Ylms in Steinhardt

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -58,15 +61,19 @@ class Steinhardt
-                        bool wl_normalize = false)
-        : m_l(l), m_num_ms(2 * l + 1), m_average(average), m_wl(wl), m_weighted(weighted),
-          m_wl_normalize(wl_normalize), m_qlm_local(2 * l + 1)
+    explicit Steinhardt(const std::vector<unsigned int>& ls, bool average = false, bool wl = false,
+                        bool weighted = false, bool wl_normalize = false)
```
Define a delegating constructor with the old signature that wraps `l` in an initializer list and forwards everything else along. That will make things easier for users intending to continue using this code the old way, especially using the C++ API.

vyasr commented on cpp/order/SolidLiquid.cc:
```diff
@@ -10,7 +10,7 @@ namespace freud { namespace order {
 SolidLiquid::SolidLiquid(unsigned int l, float q_threshold, unsigned int solid_threshold, bool normalize_q)
     : m_l(l), m_num_ms(2 * l + 1), m_q_threshold(q_threshold), m_solid_threshold(solid_threshold),
-      m_normalize_q(normalize_q), m_steinhardt(l), m_cluster()
+      m_normalize_q(normalize_q), m_steinhardt({l}), m_cluster()
```
If you define a delegating constructor as I proposed this change can be removed.

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -48,6 +49,8 @@ namespace freud { namespace order {
  * will be normalized.
  *
+ * We compute the Steinhardt order parameter for one or more l specified in m_ls.
+ *
```
I don't think this is needed, the parameter is documented (probably should rewrite that to clearly emphasize that it's one or more values).

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -58,15 +61,19 @@ class Steinhardt
     /*! Constructor for Steinhardt analysis class.
-     *  \param l Spherical harmonic number l.
-     *           Must be a positive number.
+     *  \param l Spherical harmonic numbers l. Must be non-negative integers.
```
```suggestion
     *  \param l One or more values of the spherical harmonic parameter `l`. Must be non-negative integers.
```

vyasr commented on cpp/order/Steinhardt.h:
```diff
@@ -137,16 +144,18 @@ class Steinhardt
+    //  std::vector<std::vector<std::complex<float> > > with values for m = 0, 1, ..., l, -1, ..., -l
+    //  for all l in m_ls
+    void computeYlm(fsph::PointSPHEvaluator<float>& sph_eval, const float theta, const float phi,
+                    std::vector<std::vector<std::complex<float>>>& Ylms) const;
```
Use the `YlmsType` once it's moved.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -4,55 +4,63 @@
 
 namespace freud { namespace order {
 
+    using YlmsType = std::vector<std::vector<std::complex<float>>>;
```
Move this to the header so you can also use it there.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -90,15 +101,35 @@ void Steinhardt::compute(const freud::locality::NeighborList* nlist,
         [=](size_t i, const std::shared_ptr<freud::locality::NeighborPerPointIterator>& ppiter) {
             float total_weight(0);
             const vec3<float> ref((*points)[i]);
+            // Construct PointSPHEvaluator outside loop since the construction is costly.
+            auto max_l = *std::max_element(m_ls.begin(), m_ls.end());
+            fsph::PointSPHEvaluator<float> sph_eval(max_l);
```
It's worth double-checking this in case I'm wrong (haven't looked at fsph much in a while), but IIRC the evaluator is stateless and constructing the iterators is entirely stateless, so this logic can be removed from the lambda entirely and the evaluator captured by reference in the lambda. Probably won't have any meaningful performance impact relative to doing this in parallel, but it's cleaner if you can do this once and is a slightly reduction in the code's memory footprint.

b-butler replied:
I could be wrong, but since the `PointSPHEvaluator` precomputes quantities based on theta and phi we wouldn't want to capture by reference since this would lead to race conditions when `compute` is called with the current thread's theta and phi.

Edit: add permalink https://github.com/glotzerlab/fsph/blob/11d4e640f14d25e74212d9bd1ca784e39206ba7c/src/spherical_harmonics.hpp#L181-L195

vyasr replied:
You're right, ignore this suggestion.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -122,32 +153,49 @@ void Steinhardt::baseCompute(const freud::locality::NeighborList* nlist,
+                    auto& Ylm = Ylms[l_index];
+                    // Get the initial index and iterate using ++ for faster iteration
+                    // Profiling showed using operator() to slow the code significantly.
+                    const size_t index = qlmi.getIndex({i, 0});
+                    for (size_t k = 0; k < m_num_ms[l_index]; ++k)
+                    {
+                        qlmi[index + k] += weight * Ylm[k];
+                    }
```
```suggestion
                    const size_t index = qlmi.getIndex({i, 0});
                    const size_t max_index = index + m_num_ms[l_index];
                    for (size_t k = index; k < max_index; ++k)
                    {
                        qlmi[k] += weight * Ylm[k];
                    }
```

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -122,32 +153,49 @@ void Steinhardt::baseCompute(const freud::locality::NeighborList* nlist,
                 {
-                    m_qlm_local.local()[k] += m_qlmi[index] / float(m_Np);
+                    // Cache the index for efficiency.
+                    qlmi[index + k] /= total_weight;
```
You can make the same change that I made above (starting the loop index `k` at `index`) to avoid having to shift the index inside this loop. That'll help you more here since it's happening multiple times.

b-butler replied:
Since `qlm_local` is indexed by the current `k` I would have to have two index variables still, but I can do this.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,80 +236,117 @@ void Steinhardt::computeAve(const freud::locality::NeighborList* nlist,
+                        auto& qlmi = m_qlmi[l_index];
+                        const auto ave_index = qlmiAve.getIndex({i, 0});
+                        const auto nb2_index = qlmi.getIndex({nb2.point_idx, 0});
+                        for (size_t k = 0; k < m_num_ms[l_index]; ++k)
```
Loop over `[ave_index, ave_index + m_num_ms[l_index)`.

bdice replied:
I think that might not work here -- we also need to index by `nb2_index + k`.

vyasr replied:
You're right, I'm resolving this suggestion.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,80 +236,117 @@ void Steinhardt::computeAve(const freud::locality::NeighborList* nlist,
+                const size_t index = qlmiAve.getIndex({i, 0});
+                const size_t qliAve_index = qliAve_i_start + l_index;
+
+                for (size_t k = 0; k < m_num_ms[l_index]; ++k)
```
Shift loop indices.

vyasr commented on cpp/order/Steinhardt.cc:
```diff
@@ -184,80 +236,117 @@ void Steinhardt::computeAve(const freud::locality::NeighborList* nlist,
+                const auto l = m_ls[l_index];
+                const auto& source_l = source[l_index];
+
+                const auto normalizationfactor = static_cast<float>(4.0 * M_PI / m_num_ms[l_index]);
```
Considering how often we compute this normalization factor, perhaps it would be worth defining a compile-time evaluated template function `getNormalizationFactor<num_ms>()` encoding the values for the supported range of `l` values? Currently we're limited anyway by the values of the Wigner3j coefficients we hard code (I plan to replace that eventually), so you could match that with a compile-time generation of the normalization. I'm happy to take that on myself at a later point if you want to punt, though.

vyasr replied:
Punting

vyasr commented on freud/order.pyx:
```diff
@@ -598,41 +605,59 @@ cdef class Steinhardt(_PairCompute):
     def l(self):  # noqa: E743
         """unsigned int: Spherical harmonic quantum number l."""
-        return self.thisptr.getL()
+        # list conversion is necessary as otherwise CI Cython complains about
+        # compiling the below expression with two different types.
```
:(

vyasr reviewed (changes requested):
This looks good overall. I'll defer to you on performance changes, I trust that you and Bradley have tested sufficiently. Regarding the ManagedArray discussion, the thing to keep in mind is that conceptually `ManagedArray` is really a low-level RAII wrapper around a pointer to a contiguous memory buffer whose ownership can be transferred. `std::vector` should be used more like a list, whereas `ManagedArray` should be like a numpy array. Since `l` values are arbitrary and not necessarily sequential, a `std::vector<ManagedArray>` makes sense. The expectation is that Python users acquiring `ManagedArray` variables expect that those variables should be zero-copy but remain unchanged if a recompute is triggered. Since it looks like you're replacing all the `ManagedArray` data in the vector on recompute, I think you'll be fine. I'll double-check this on my final pass though since I was a bit cursory with checking that this time around.

vyasr committed:
Merge branch 'master' into feature/multi-sph-l

b-butler committed:
Simplify for loop indexing

Co-authored-by: Vyas Ramasubramani <vramasub@umich.edu>

b-butler committed:
Improve code commment

Co-authored-by: Vyas Ramasubramani <vramasub@umich.edu>

b-butler committed:
Move YlmTypes to header to use in method declaration

b-butler committed:
Add more const

b-butler committed:
Add delegating constructor for Steinhardt for single l

b-butler committed:
Revert "Simplify for loop indexing"

This reverts commit 6d91a73af67a7ac92f60204a6bf1a35671c3f9a8.
The loop simplifaction ignored that the index variable was shifted
differently depending on the array.

b-butler commented:
@vyasr most of your comments on loops I disagree with shifting the loop bounds would simplify the code, since the different arrays are indexed differently. In all cases at least two four loop variables would need to be defined and incremented. If you prefer multiple indices to be created in a for loop, then I will make the change.

b-butler committed:
Remove unnecessary comment

bdice commented on cpp/order/SolidLiquid.cc:
```diff
@@ -29,7 +29,9 @@ void SolidLiquid::compute(const freud::locality::NeighborList* nlist,
-    const auto& qlm = m_steinhardt.getQlm();
+    const auto& qlm = m_steinhardt.getQlm()[0];
+    // Even though this is a 2-D array since Steinhardt can compute multiple l the second dimension is always
+    // 1 since we only allow one l for SolidLiquid.
```
Put the comment above the line to which it applies. Also this comment can be much simpler.
```suggestion
    // SolidLiquid only has one l value so we index the 2D array from Steinhardt.
    const auto& qlm = m_steinhardt.getQlm()[0];
```

bdice commented on cpp/order/Steinhardt.cc:
```diff
@@ -4,55 +4,62 @@
 
 namespace freud { namespace order {
 
+
```
```suggestion
```

bdice reviewed (commented):
A couple more comments.

bdice committed:
Remove extra newline.

vyasr commented:
> @vyasr most of your comments on loops I disagree with shifting the loop bounds would simplify the code, since the different arrays are indexed differently. In all cases at least two four loop variables would need to be defined and incremented. If you prefer multiple indices to be created in a for loop, then I will make the change.

Caught up with @b-butler briefly offline. I agree that in many cases you still need to maintain multiple indices, so just shifting the bounds may not be so helpful. I am a little hesitant to rely on the compiler to optimize out the multiple computations of the loop index, so I have a slight preference for computing it once and storing it to make the compiler's job easier, but aside from that I don't think any changes are necessary on the indexing front. I'll leave it to @b-butler to make the call there, then please re-request my review and we can wrap this up.

vyasr committed:
Merge branch 'master' into feature/multi-sph-l

bdice commented:
@b-butler Let me know if you want help on anything here, it'd be helpful for my research code to get this merged. 👍

b-butler committed:
Store index variables used multiple times

b-butler committed:
Put comment above code.

Co-authored-by: Bradley Dice <bdice@bradleydice.com>

vyasr committed:
Merge branch 'master' into feature/multi-sph-l

vyasr reviewed (approved):
Looks good, we can make further changes in later iterations if we need to.
